### ISSUE TITLE
Codegen : Complex Implementation

### ISSUE THREAD

**Issue by SatyaPrakashDwibedi** [sympy/sympy#12453]
consider the following code snippet
```
>>> [(c,d),(e,f)]=codegen(("f",(x+y*I)**2),"C",header=False,empty=False)
>>> print d
#include "f.h"
#include <math.h>
double f(double x, double y) {
   double f_result;
   f_result = pow(x + I*y, 2);
   return f_result;
}
```
should generate:
```
#include "f.h"
#include <complex.h>
double complex f(double x,double y){
     double complex f_result;
     f_result = cpow( x + I*y , 2);
    return f_result;
}
```
i.e we have to add complex library for calculation involving complex symbols in c.

**Comment by asmeurer** [comment]
The type of the final function should probably be specified as an argument to codegen.

**Comment by SatyaPrakashDwibedi** [comment]
Then how do you suggest to implement inbuilt complex functions  like `cpow()` as stated above

**Comment by asmeurer** [comment]
I think it should still use complex.h, but the types should be user specified as a keyword argument to codegen. If the user specifies dtype="double complex" then it will add the correct header files and code print cpow and so on. For instance, not all complex functions will have an explicit `I` in them for the codegen function to automatically deduce that the expression is complex.

**Comment by SatyaPrakashDwibedi** [comment]
so you are suggesting that we should add an attribute to codegen function as input i.e `dtype` and set it as 
`double` by default.

**Comment by SatyaPrakashDwibedi** [comment]
so you want to add a dtypes argument to codegen function right??
I think it will be awesome because using double everywhere even in case of integers takes more space in C.

**Comment by asmeurer** [comment]
Yes. Doing types correctly requires a bit more thought, because different variables might have different types. And we may at some point want to add the ability to do type inferences (so for instance, if we are given an expression, we can determine its type from the types of the variables used in it).

**Comment by SatyaPrakashDwibedi** [comment]
@asmeurer yes this will make user less worried about the datatypes .
There we can check the symbol types and print specific datatypes which will be memory efficient and print corresponding datatype or we can add a dictionary mapping from sympy symbol types to C data type
for e.g.
```
{'complex':'double , complex','int':'int'}
```
so that when that type of symbol is passed it will print the correct datatype in C.
Reg. addition of `complex.h` header file
we can check if symbol type is complex then only we can proceed to add `complex.h` header file else adding it every where will be redundant and in efficient.

**Comment by ChrisPattison** [comment]
I think it's worthwhile to mention in the docs that sympy currently does handle codegen for complex valued calculations well. It was quite a surprise to me to find that C allows complex double to be implicitly casted to double
